src/c++/perf_analyzer/genai-perf/tests/test_cmd_builder.py

```diff
@@ -0,0 +1,88 @@
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import pytest
+from genai_perf import parser
+from genai_perf.constants import DEFAULT_GRPC_URL
+from genai_perf.wrapper import Profiler
+
+
+class TestCmdBuilder:
+    @pytest.mark.parametrize(
+        "arg",
+        [
+            ([]),
+            (["-u", "testurl:1000"]),
+            (["--url", "testurl:1000"]),
+        ],
+    )
+    def test_url_exactly_once(self, monkeypatch, arg):
+        args = ["genai-perf", "-m", "test_model"] + arg
+        monkeypatch.setattr("sys.argv", args)
+        args, extra_args = parser.parse_args()
+        cmd_string = Profiler.build_cmd(args, extra_args)
+
+        number_of_url_args = cmd_string.count(" -u ") + cmd_string.count(" --url ")
+        assert number_of_url_args == 1
+
+    @pytest.mark.parametrize(
+        "arg",
+        [
+            (["--output-format", "openai_chat_completions"]),
+            (["--output-format", "openai_completions"]),
+            (["--output-format", "trtllm"]),
+            (["--output-format", "vllm"]),
+        ],
+    )
+    def test_service_triton(self, monkeypatch, arg):
+        args = ["genai-perf", "-m", "test_model", "--service-kind", "triton"] + arg
+        monkeypatch.setattr("sys.argv", args)
+        args, extra_args = parser.parse_args()
+        cmd_string = Profiler.build_cmd(args, extra_args)
+
+        assert cmd_string.count(" -i grpc") == 1
+        assert cmd_string.count(" --streaming") == 1
+        assert cmd_string.count(f"-u {DEFAULT_GRPC_URL}") == 1
+
+        if arg == "trtllm":
+            assert cmd_string.count("--shape max_tokens:1") == 1
+            assert cmd_string.count("--shape text_input:1") == 1
+
+    def test_service_openai(self, monkeypatch):
+        args = [
+            "genai-perf",
+            "-m",
+            "test_model",
+            "--service-kind",
+            "openai",
+            "--endpoint",
+            "v1/completions",
+        ]
+        monkeypatch.setattr("sys.argv", args)
+        args, extra_args = parser.parse_args()
+        cmd_string = Profiler.build_cmd(args, extra_args)
+
+        assert cmd_string.count(" -i http") == 1
```

Comment:
I think a brief comment explaining this check would be good for future readers

Reply:
How's the comment look? I looked at what we're doing in [add_protocol_args](https://github.com/triton-inference-server/client/blob/81d0925e31bbcc315965f831a906a70a634d60f9/src/c%2B%2B/perf_analyzer/genai-perf/genai_perf/wrapper.py#L39-L49) and tested the current logic to ensure it doesn't change without breaking a test.

Tagged Elias and Tim as reviewers, since they are more familiar with the reason for the logic for each service-kind, so if it needs to be more malleable or stricter, they can let me know for this PR.

Reply:
I think I was expecting something like `"OpenAI service-kind implicitly expects HTTP protocol under the hood"`

Separately, a related (bonus) power-user test/check is probably that something like `genai-perf --service-kind openai ... -- -i grpc` would gracefully fail? But understood if power-user combinations are not in scope to test.

Reply:
I think my pushback here would be that then we're commenting and coding the same things. We'd need to add a comment to the previous test as well for Triton expecting GRPC, streaming, etc. My hope is that the tests are self-commenting, and I think they are. You'd now that we're testing for these args... then you'd have to look at the implementation or try running the commands to know why we only support those args.

If we wanted to add additional information about the conditional logic, I think the better place would be in the implementation code (e.g. "OpenAI service-kind implicitly expects HTTP protocol. [TMA-XXXX] Add GRPC support."). That said, I just tried running the power user command you just mentioned, and perf_analyzer does have good error messages returned, so in that way there is already some commenting (that would be missed by someone just looking at the GenAi-PA code, but would be found if they try running with a different mode):
![image](https://github.com/triton-inference-server/client/assets/58150256/f2d5a862-5615-4c9c-bce5-745c9f87968f)

Yeah, Elias mentioned that decision to include the power user mode for those who want greater control/flexibility, but there is no intent to add guardrails around it beyond printing that passthrough args have been detected and listing them. It'd be too open-ended to test.

Reply:
Thank you for your expedient, detailed reviews by the way, Ryan! I always feel like I learn something.

Feel free to push back on my pushback. If you think there's a better way to comment the code (or use self-commenting code), we could adapt.

Reply:
> I think my pushback here would be that then we're commenting and coding the same things. We'd need to add a comment to the previous test as well for Triton expecting GRPC, streaming, etc. My hope is that the tests are self-commenting, and I think they are.

Understood my example comment is a little redundant based on _what_ the test is doing, agree with you there upon re-reading it. My main point was _why_ the test was checking that. It seemed like an unintuitive implementation detail you'd have to go read the implementation to find out about, and as a reviewer I didn't want to go read the implementation to find that detail, so just shared my thoughts.

I don't have a better suggested comment off the top of my head, and it's definitely a nit -- I'm happy with how it currently is :+1:

Reply:
I think a lot of what you are saying Ryan will evolve in the tool as we add more and more error checking.
For now, as i understand it, if a user does OpenAI + grpc, you are not following the OpenAI api anymore. So we assume if you want openai, you dont need to specify http. We set it for you to make the command you have to enter on the command line overall less verbose.